rfcs/0009-code-review.md

```diff
@@ -0,0 +1,182 @@
+---
+created-on: 2021-04-27
+freshness: 2021-04-27
+owner: @dj2
+status: proposal
+tags:
+  - software development
+  - practices
+---
+
+[La version française suit.](#résumé)
+
+# Code Review
+
+## Summary
+Code review is an integral part of how we work at CDS. Having your code
+reviewed provides an extra layer of experience and guidance. It's easy to
+get lost in the forest and miss potential complications and
+simplifications.  Providing reviews gives the opportunity to understand
+and shape the system as a whole.
+
+The process on both sides allows our developers to learn from each other.
+As much can be learned by the reviewer as the reviewee.
+
+This RFC lays out considerations for reviewing code at CDS.
+
+## Motivation
+A shared set of considerations helps ease both the burden of being
+reviewed and that of being the reviewer. Code review is inherently
+subjective, different reviewers focus on different things. This can make
+the review system seem opaque. Shared considerations help to ease that
+subjectivity by giving guide rails to make reviews more consistent.
+
+## Proposal
+Below is a list of considerations these are divided into principles and
+technical consideration. Focus on the principles and then lead into the
+technical considerations.
+
+### General guidelines
+ - Code reviews are required at CDS.
+   - Each review should have a technical reviewer.
+   - Add Design, Research, Product Management, Service Owner, as needed.
+ - Tend towards questions rather then statements.
+   - The reviewer may not understand all the considerations a reviewee
+     put into the code.
+   - The reviewee may not understand the reviewers past experience with
+     the code.
+ - [Be nice](https://qz.com/work/625870/after-years-of-intensive-analysis-google-discovers-the-key-to-good-teamwork-is-being-nice/)
+   - Read the CDS [Code of Conduct](https://github.com/cds-snc/.github/blob/master/CODE_OF_CONDUCT.md)
+ - Escalate if the review is blocked.
+   - The goal of the review is that the code looks good to the reviewer.
```

Comment:
What is 'good'?  Wondering if we need to have clearer common understanding.  For example:
- Is 'good' tied to the reviewers underlying knowledge of the programming language used and/or the project itself?
- Is a 'hacky' solution good because it meets an immediate need vs introducing tech debt and future work refactoring?

Reply:
That's a really good question and .... I think it depends?

To me, the answer is very context dependant. Sometimes it's worthwhile taking on tech debt to accomplish some goal. You need to know you're taking on the debt and have some plans around containing it, but taking it on maybe the right call. So, a hacky solution maybe the correct solution for now.

So, I think for this 'good' means in the context of the timelines, goals and constraints of the PR.  It's also tied to the reviewers abilities. My ability to review 'good' python code is small, I can do it but I'm not a python expert (or intermediate) so my good is a lot different then other folks. My good would be understanding as a non-python dev, comments, structure, but I'd probably also request a python person review the code as well.

Thoughts?

Reply:
I agree the sentence is too ambiguous. Could we use the Technical Considerations as a north star goal for any code review? For e.g. _The goal of the review is that the reviewers ensure the code follows the technical considerations defined in this RFC._

Reply:
Better?